src/v/coproc/event_handler.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "coproc/event_handler.h"
+
+#include "coproc/ntp_context.h"
+
+namespace coproc::wasm {
+
+async_event_handler::async_event_handler(ss::sharded<pacemaker>& pacemaker)
+  : _dispatcher(pacemaker, _abort_source) {}
+
+ss::future<> async_event_handler::start() { co_return; }
+
+ss::future<> async_event_handler::stop() {
+    _abort_source.request_abort();
+    co_return;
```

Comment:
the general pattern is that when stopping an abort is requested and then a future is returned which resolves when whatever was being aborted also has stopped. It doesn't mean this is wrong, but it looks suspicious. Why request an abort at this level if this level doesn't wait on anything to stop? Perhaps you should be passing in a _reference_ to an abort source managed at a higher level.

Reply:
In last version I used ref to _abort_source from event_listener. 
But creating async_handler looked like:
```
async_handler = std::make_unique(event_listener->get_abort_source);
event_listener->register(async_handler)
```

For me it is more clearly if I do not pass ref to abort_source in async_handler

Reply:
I think that you just need to use gate here. Other things that can be executed asynchronously should be wrapped with gate and here you can return gate.close()

Reply:
I need abort_source because it is used in 
```
ss::future<std::optional<coproc::supervisor_client_protocol>>
script_dispatcher::get_client() {
    model::timeout_clock::duration dur = 1s;
    while (true) {
        if (_abort_source.abort_requested()) {
            co_return std::nullopt;
        }
        auto timeout = model::timeout_clock::now() + 100ms;
        auto transport = co_await _transport.get_connected(timeout);
        if (!transport) {
            vlog(
              coproclog.error,
              "script_dispatcher failed to aquire a connection to the wasm "
              "engine (to deploy or remove a script), retrying... ");
            co_await ss::sleep(dur);
            dur = std::min(model::timeout_clock::duration(10s), dur * 2);
        } else {
            co_return supervisor_client_protocol(*transport.value());
        }
    }
}
```

for stop iteration.

Reply:
What I'm trying to say is that async_event_handler is async, and many of its methods return futures. When you're calling `stop` method it should wait for all async things that might be in-flight. This is usually done by guarding all methods that return futures with the gate. And the stop methods can call `request_abort` and then wait on `gate.close`.

Reply:
Added gate, but I still think about abort_source